### PR TITLE
Roles refactoring: Add new project_member table, update getter/setter…

### PR DESCRIPTION
**Pull request by MarcelGeo**
Resolves https://github.com/MerginMaps/server-private/issues/2586

### CODE REVIEW

**Review by MarcelGeo — COMMENTED** [review]
migration is not running without intarray extension in postgres.
possible alternative would be use EXCEPT operator

```sql
SELECT DISTINCT project_id, user_id, role FROM (SELECT
                        project_id AS project_id,
                        unnest(owners) AS user_id,
                        'owner' AS role
                FROM project_access
UNION
SELECT
    project_id,
    unnest(writer_ids) AS user_id,
    'writer' AS role
FROM (
    SELECT
        project_id,
        array(SELECT UNNEST(writers) EXCEPT SELECT UNNEST(owners)) AS writer_ids
    FROM project_access
) AS subquery
UNION
SELECT
    project_id,
    unnest(writer_ids) AS user_id,
    'writer' AS role
FROM (
    SELECT
        project_id,
        array(SELECT UNNEST(editors) EXCEPT SELECT UNNEST(array_cat(writers, owners))) AS writer_ids
    FROM project_access
) AS subquery) unioned
;
```

Another problem shoukd be with results of query, where multiple non unique users ids is returned. 


do we want to use intarray extension?

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            ondelete="CASCADE",
+        ),
+        sa.ForeignKeyConstraint(
+            ["user_id"],
+            ["user.id"],
+            name=op.f("fk_project_member_user_id_user"),
+            ondelete="CASCADE",
+        ),
+        sa.PrimaryKeyConstraint(
+            "project_id", "user_id", name=op.f("pk_project_member")
```
Comment (mostlyAtNight) [inline server/migrations/community/d02961c7416c_add_project_member_table.py:46]: Is it therefore true that there'll only ever be one row in this table, regardless of access level, for a given user on a given project?

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+
+def downgrade():
+    op.create_table(
+        "project_access",
+        sa.Column("public", sa.BOOLEAN(), autoincrement=False, nullable=True),
+        sa.Column(
+            "owners",
+            postgresql.ARRAY(sa.INTEGER()),
+            server_default=sa.text("'{}'::integer[]"),
```
Comment (mostlyAtNight) [inline server/migrations/community/d02961c7416c_add_project_member_table.py:66]: This is different to the schema we run right now in production (we don't use a default)

**Review by mostlyAtNight — COMMENTED** [review]
Hi guys, I've reviewed the migration script thoroughly but have not looked at any other part of the MR.

Some other thoughts:
- we should ensure an upgrade/downgrade cycle results in exactly the same states for the tables affected 
- do we need to grant any permissions on new tables, or these will be taken care of automaticlaly with default schema permissions?

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            name=op.f("fk_project_member_user_id_user"),
+            ondelete="CASCADE",
+        ),
+        sa.PrimaryKeyConstraint(
+            "project_id", "user_id", name=op.f("pk_project_member")
+        ),
+    )
+
+    op.add_column("project", sa.Column("public", sa.Boolean(), nullable=True))
+    op.create_index(op.f("ix_project_public"), "project", ["public"], unique=False)
```
[the remark this answers is not in the document]
Comment (mostlyAtNight) [reply server/migrations/community/d02961c7416c_add_project_member_table.py:51]: Feel free to ignore this

File: server/migrations/community/d02961c7416c_add_project_member_table.py (hunk 2 above)
Comment (mostlyAtNight) [reply server/migrations/community/d02961c7416c_add_project_member_table.py:66]: Discussed and this should not matter

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                FROM project_access
+                UNION
+                SELECT
+                        project_id AS project_id,
+                        unnest(readers - editors - writers - owners) AS user_id,
+                        'reader' AS role
+                FROM project_access
+            )
+            INSERT INTO project_member (project_id, user_id, role)
+            SELECT DISTINCT project_id, user_id, role::project_role FROM members;
```
[the remark this answers is not in the document]
Comment (mostlyAtNight) [reply server/migrations/community/d02961c7416c_add_project_member_table.py:175]: It seems we are happy with this and will confirm with our upgrade/downgrade consistency check

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                        project_member
+                    GROUP BY project_id, role
+                )
+                SELECT
+                    o.project_id,
+                    o.users_ids AS owners,
+                    o.users_ids || w.users_ids AS writers,
+                    o.users_ids || w.users_ids || e.users_ids AS editors,
+                    o.users_ids || w.users_ids || e.users_ids || r.users_ids AS readers
+                FROM (SELECT * FROM agg WHERE role = 'owner') AS o
```
[the remark this answers is not in the document]
Comment (mostlyAtNight) [reply server/migrations/community/d02961c7416c_add_project_member_table.py:214]: Let's tackle this before doing the upgrade/downgrade check

File: server/mergin/auth/schemas.py [hunk 6]
```diff
@@ -33,17 +33,16 @@ def get_disk_usage(self, obj):
...
         # DEPRECATED functionality - kept for the backward-compatibility
-        from ..sync.models import Project, ProjectAccess
+        from ..sync.models import ProjectUser, Project
 
         ws = current_app.ws_handler.get_by_name(obj.user.username)
         if ws:
             projects_count = (
-                Project.query.filter(Project.creator_id == obj.user.id)
+                Project.query.join(ProjectUser)
+                .filter(Project.creator_id == obj.user.id)
```
Comment (MarcelGeo) [inline server/mergin/auth/schemas.py:42]: creator_id looks like new condition here. Do we need this check?

File: server/mergin/sync/models.py [hunk 7]
```diff
@@ -253,24 +257,43 @@ def delete(self, removed_by: int = None):
...
+    EDITOR = "editor"
+    WRITER = "writer"
+    OWNER = "owner"
 
-        https://docs.python.org/3/library/enum.html#enum.EnumType.__members__
-        """
+    def __ge__(self, other):
+        """Compare project roles"""
         members = list(ProjectRole.__members__)
-        if members.index(self.name) < members.index(other.name):
```
Comment (MarcelGeo) [inline server/mergin/sync/models.py:270]: Are we sure with this change of order?

**Review by MarcelGeo — COMMENTED** [review]
Downgrade, upgrade working for my local data. I also checked if everything is ok with data.

File: server/migrations/community/d02961c7416c_add_project_member_table.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        ),
+    )
+
+    op.add_column("project", sa.Column("public", sa.Boolean(), nullable=True))
+    op.create_index(op.f("ix_project_public"), "project", ["public"], unique=False)
+
+    data_upgrade()
+
+    op.drop_table("project_access")
+    conn.execute(sa.text("DROP EXTENSION IF EXISTS intarray;"))
```
[the remark this answers is not in the document]
Comment (MarcelGeo) [reply server/migrations/community/d02961c7416c_add_project_member_table.py:56]: Then just create extension without dropping. I thinks there is not issue to have it in other deployments

File: server/mergin/auth/schemas.py (hunk 6 above)
Comment (varmar05) [reply server/mergin/auth/schemas.py:42]: creator_id was there before in filter, but in fact it is some legacy stuff for backward compatibility

File: server/migrations/community/d02961c7416c_add_project_member_table.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (varmar05) [reply server/migrations/community/d02961c7416c_add_project_member_table.py]: done

File: server/migrations/community/d02961c7416c_add_project_member_table.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (varmar05) [reply server/migrations/community/d02961c7416c_add_project_member_table.py]: adjusted to count all projects - as it was causing inconsistencies

**Review by mostlyAtNight — APPROVED** [review]
Looks all good from my side

File: server/mergin/auth/schemas.py (hunk 6 above)
Comment (MarcelGeo) [reply server/mergin/auth/schemas.py:42]: ok ... this looks like legacy method :)

**Pull request merged**